mm2src/coins/z_coin/z_htlc.rs

```diff
@@ -82,3 +101,141 @@ pub async fn z_send_htlc(
+        hash,
+        checksum_type: coin.utxo_arc.conf.checksum_type,
+    };
+
+    let amount_sat = sat_from_big_decimal(&amount, coin.utxo_arc.decimals).expect("temporary code");
+    let amount = big_decimal_from_sat_unsigned(amount_sat, coin.utxo_arc.decimals);
+
+    let address = htlc_address.to_string();
+    if let UtxoRpcClientEnum::Native(native) = coin.rpc_client() {
+        native.import_address(&address, &address, false).compat().await.unwrap();
+    }
+
+    let send_item = ZSendManyItem {
+        amount,
+        op_return: Some(payment_script.to_vec().into()),
+        address,
+    };
+
+    let op_id = coin
+        .z_rpc()
+        .z_send_many(&coin.z_fields.z_addr_encoded, vec![send_item])
+        .compat()
+        .await?;
+
+    loop {
+        let operation_statuses = coin.z_rpc().z_get_send_many_status(&[&op_id]).compat().await?;
+
+        match operation_statuses.first() {
+            Some(ZOperationStatus::Executing { .. }) | Some(ZOperationStatus::Queued { .. }) => {
+                Timer::sleep(1.).await;
+                continue;
+            },
+            Some(ZOperationStatus::Failed { .. }) => {
+                break Err(MmError::new(ZSendHtlcError::ZOperationFailed(operation_statuses)));
+            },
+            Some(ZOperationStatus::Success { result, .. }) => {
+                let tx_bytes = coin
+                    .rpc_client()
+                    .get_transaction_bytes(result.txid.clone())
+                    .compat()
+                    .await?;
+                let tx: UtxoTx = deserialize(tx_bytes.0.as_slice()).expect("rpc returns valid tx bytes");
+
+                coin.rpc_client()
+                    .wait_for_confirmations(&tx, 1, false, now_ms() / 1000 + 120, 1)
+                    .compat()
+                    .await
+                    .unwrap();
+                break Ok((tx, payment_script));
+            },
+            None => break Err(MmError::new(ZSendHtlcError::ZOperationStatusesEmpty)),
+        }
+    }
+}
+
+#[derive(Debug, Display)]
+#[allow(clippy::large_enum_variant)]
+pub enum ZP2SHSpendError {
+    ZTxBuilderError(ZTxBuilderError),
+    Rpc(UtxoRpcError),
+}
+
+impl From<ZTxBuilderError> for ZP2SHSpendError {
+    fn from(tx_builder: ZTxBuilderError) -> ZP2SHSpendError { ZP2SHSpendError::ZTxBuilderError(tx_builder) }
+}
+
+impl From<UtxoRpcError> for ZP2SHSpendError {
+    fn from(rpc: UtxoRpcError) -> ZP2SHSpendError { ZP2SHSpendError::Rpc(rpc) }
+}
+
+/// Spends P2SH output 0 to the coin's z_addr
+pub async fn z_p2sh_spend(
+    coin: &ZCoin,
+    p2sh_tx: UtxoTx,
+    tx_locktime: u32,
+    input_sequence: u32,
+    redeem_script: Script,
+    script_data: Script,
+) -> Result<UtxoTx, MmError<ZP2SHSpendError>> {
+    let current_block = coin.utxo_arc.rpc_client.get_block_count().compat().await? as u32;
+    let mut tx_builder = ZTxBuilder::new(consensus::MAIN_NETWORK, current_block.into());
```

Comment:
Only one question, do we need to support `TEST_NETWORK` for zcoin?

Reply:
Yes, for now, it's fine to stick with mainnet only and I plan to make the implementation more generic.